File: TarsEngineFSharp/Result.fs
namespace TarsEngineFSharp

/// <summary>
/// Discriminated union for representing a result that can be either a success or a failure
/// </summary>
type Result<'TSuccess, 'TFailure> =
    /// <summary>
    /// Represents a successful result
    /// </summary>
    | Success of 'TSuccess
    /// <summary>
    /// Represents a failed result
    /// </summary>
    | Failure of 'TFailure

/// <summary>
/// Module containing helper functions for Result
/// </summary>
module Result =
    /// <summary>
    /// Create a success result
    /// </summary>
    /// <param name="value">The success value</param>
    /// <returns>A success result</returns>
    let success value = Success value

    /// <summary>
    /// Create a failure result
    /// </summary>
    /// <param name="error">The error value</param>
    /// <returns>A failure result</returns>
    let failure error = Failure error

    /// <summary>
    /// Apply a function to the success value of a result
    /// </summary>
    /// <param name="f">The function to apply</param>
    /// <param name="result">The result</param>
    /// <returns>A new result with the function applied to the success value</returns>
    let map f result =
        match result with
        | Success value -> Success (f value)
        | Failure error -> Failure error

    /// <summary>
    /// Apply a function to the failure value of a result
    /// </summary>
    /// <param name="f">The function to apply</param>
    /// <param name="result">The result</param>
    /// <returns>A new result with the function applied to the failure value</returns>
    let mapError f result =
        match result with
        | Success value -> Success value
        | Failure error -> Failure (f error)

    /// <summary>
    /// Apply a function that returns a result to the success value of a result
    /// </summary>
    /// <param name="f">The function to apply</param>
    /// <param name="result">The result</param>
    /// <returns>A new result with the function applied to the success value</returns>
    let bind f result =
        match result with
        | Success value -> f value
        | Failure error -> Failure error

    /// <summary>
    /// Apply a function to both the success and failure values of a result
    /// </summary>
    /// <param name="onSuccess">The function to apply to the success value</param>
    /// <param name="onFailure">The function to apply to the failure value</param>
    /// <param name="result">The result</param>
    /// <returns>The result of applying the appropriate function</returns>
    let either onSuccess onFailure result =
        match result with
        | Success value -> onSuccess value
        | Failure error -> onFailure error

    /// <summary>
    /// Get the success value or a default value if the result is a failure
    /// </summary>
    /// <param name="defaultValue">The default value</param>
    /// <param name="result">The result</param>
    /// <returns>The success value or the default value</returns>
    let defaultValue defaultValue result =
        match result with
        | Success value -> value
        | Failure _ -> defaultValue

    /// <summary>
    /// Get the success value or apply a function to the failure value
    /// </summary>
    /// <param name="defaultFn">The function to apply to the failure value</param>
    /// <param name="result">The result</param>
    /// <returns>The success value or the result of applying the function to the failure value</returns>
    let defaultWith defaultFn result =
        match result with
        | Success value -> value
        | Failure error -> defaultFn error

    /// <summary>
    /// Try to execute a function and return a result
    /// </summary>
    /// <param name="f">The function to execute</param>
    /// <returns>A success result with the function result or a failure result with the exception</returns>
    let tryWith f =
        try
            Success (f())
        with
        | ex -> Failure ex

File: v1/parked_csharp/Legacy_CSharp_Projects/DuplicationAnalyzerTests/Monads/Result.cs
using System;
using System.Threading.Tasks;

namespace DuplicationAnalyzerTests.Monads;

/// <summary>
/// Represents the result of an operation that can succeed or fail
/// </summary>
/// <typeparam name="T">The type of the result value</typeparam>
public class Result<T>
{
    /// <summary>
    /// Gets a value indicating whether the operation was successful
    /// </summary>
    public bool IsSuccess { get; }

    /// <summary>
    /// Gets the value of the result (if successful)
    /// </summary>
    public T Value { get; }

    /// <summary>
    /// Gets the error message (if failed)
    /// </summary>
    public string Error { get; }

    /// <summary>
    /// Gets the exception (if failed)
    /// </summary>
    public Exception? Exception { get; }

    private Result(bool isSuccess, T? value, string error, Exception? exception)
    {
        IsSuccess = isSuccess;
        Value = value!;
        Error = error;
        Exception = exception;
    }

    /// <summary>
    /// Creates a successful result
    /// </summary>
    /// <param name="value">The result value</param>
    /// <returns>A successful result</returns>
    public static Result<T> Success(T value) => new Result<T>(true, value, string.Empty, null);

    /// <summary>
    /// Creates a failed result
    /// </summary>
    /// <param name="error">The error message</param>
    /// <returns>A failed result</returns>
    public static Result<T> Failure(string error) => new Result<T>(false, default!, error, null);

    /// <summary>
    /// Creates a failed result
    /// </summary>
    /// <param name="exception">The exception</param>
    /// <returns>A failed result</returns>
    public static Result<T> Failure(Exception exception) => new Result<T>(false, default!, exception.Message, exception);

    /// <summary>
    /// Binds the result to another operation
    /// </summary>
    /// <typeparam name="TResult">The type of the new result</typeparam>
    /// <param name="func">The function to apply to the value</param>
    /// <returns>The new result</returns>
    public Result<TResult> Bind<TResult>(Func<T, Result<TResult>> func)
    {
        if (!IsSuccess)
        {
            return Result<TResult>.Failure(Error);
        }

        try
        {
            return func(Value);
        }
        catch (Exception ex)
        {
            return Result<TResult>.Failure(ex);
        }
    }

    /// <summary>
    /// Maps the result to a new value
    /// </summary>
    /// <typeparam name="TResult">The type of the new result</typeparam>
    /// <param name="func">The function to apply to the value</param>
    /// <returns>The new result</returns>
    public Result<TResult> Map<TResult>(Func<T, TResult> func)
    {
        if (!IsSuccess)
        {
            return Result<TResult>.Failure(Error);
        }

        try
        {
            return Result<TResult>.Success(func(Value));
        }
        catch (Exception ex)
        {
            return Result<TResult>.Failure(ex);
        }
    }

    /// <summary>
    /// Binds the result to an async operation
    /// </summary>
    /// <typeparam name="TResult">The type of the new result</typeparam>
    /// <param name="func">The async function to apply to the value</param>
    /// <returns>The new result</returns>
    public async Task<Result<TResult>> BindAsync<TResult>(Func<T, Task<Result<TResult>>> func)
    {
        if (!IsSuccess)
        {
            return Result<TResult>.Failure(Error);
        }

        try
        {
            return await func(Value);
        }
        catch (Exception ex)
        {
            return Result<TResult>.Failure(ex);
        }
    }

    /// <summary>
    /// Maps the result to a new async value
    /// </summary>
    /// <typeparam name="TResult">The type of the new result</typeparam>
    /// <param name="func">The async function to apply to the value</param>
    /// <returns>The new result</returns>
    public async Task<Result<TResult>> MapAsync<TResult>(Func<T, Task<TResult>> func)
    {
        if (!IsSuccess)
        {
            return Result<TResult>.Failure(Error);
        }

        try
        {
            return Result<TResult>.Success(await func(Value));
        }
        catch (Exception ex)
        {
            return Result<TResult>.Failure(ex);
        }
    }
}

/// <summary>
/// Static helper methods for Result monad
/// </summary>
public static class Result
{
    /// <summary>
    /// Creates a successful result
    /// </summary>
    /// <typeparam name="T">The type of the result value</typeparam>
    /// <param name="value">The result value</param>
    /// <returns>A successful result</returns>
    public static Result<T> Success<T>(T value) => Result<T>.Success(value);

    /// <summary>
    /// Creates a failed result
    /// </summary>
    /// <typeparam name="T">The type of the result value</typeparam>
    /// <param name="error">The error message</param>
    /// <returns>A failed result</returns>
    public static Result<T> Failure<T>(string error) => Result<T>.Failure(error);

    /// <summary>
    /// Creates a failed result
    /// </summary>
    /// <typeparam name="T">The type of the result value</typeparam>
    /// <param name="exception">The exception</param>
    /// <returns>A failed result</returns>
    public static Result<T> Failure<T>(Exception exception) => Result<T>.Failure(exception);

    /// <summary>
    /// Tries to execute a function and returns a result
    /// </summary>
    /// <typeparam name="T">The type of the result value</typeparam>
    /// <param name="func">The function to execute</param>
    /// <returns>The result of the function</returns>
    public static Result<T> Try<T>(Func<T> func)
    {
        try
        {
            return Success(func());
        }
        catch (Exception ex)
        {
            return Failure<T>(ex);
        }
    }

    /// <summary>
    /// Tries to execute an async function and returns a result
    /// </summary>
    /// <typeparam name="T">The type of the result value</typeparam>
    /// <param name="func">The async function to execute</param>
    /// <returns>The result of the function</returns>
    public static async Task<Result<T>> TryAsync<T>(Func<Task<T>> func)
    {
        try
        {
            return Success(await func());
        }
        catch (Exception ex)
        {
            return Failure<T>(ex);
        }
    }
}
